scripts: sbom: Add "west sbom" tool.

doki-nordic opened the pull request:
The "west sbom" tool creates Software Bill of Materials. It creates,
from provided input (an application build or the specific files),
a report containing list of licenses and files associated with them.

Co-authored-by: Marcin Jeliński <marcin.jelinski@nordicsemi.no>

doki-nordic commented:
Compliance checks complains about `Acknowledgement` vs. `Acknowledgment`.

Those are false positives errors because `Acknowledgement` comes from official SPDX license list and cannot be changed in this PR.

https://spdx.org/licenses/zlib-acknowledgement.html

mbolivar-nordic reviewed (changes requested):
This is the first I've heard of this effort. I think the commit logs do not explain why this is necessary when we already have a `west spdx` upstream. Why is upstream's tool not enough? Why create a whole new tool instead of improve the upstream tool to cover missing features? Etc. Please justify *why* this tool should be added to NCS, not just describe the tool.

doki-nordic commented:
> This is the first I've heard of this effort.

It was on the scope of next major SDK for a while. There is a discussion about it for some time now. Looks like you wasn't part of it. I will send you internal details later.

> I think the commit logs do not explain why this is necessary when we already have a `west spdx` upstream. Why is upstream's tool not enough? Why create a whole new tool instead of improve the upstream tool to cover missing features? Etc.

There are number of reasons:

1. The input.
`west spdx` reads cmake configuration files generated by cmake in the application build directory. This method does not include all input files, especially all the header files are missing. It is not possible to get the header files with this method, so something new had to be written. This part of the `west spdx` script takes most of the source code and it is not simple. We need all files, so it has be be rewritten if we wanted to adjust upstream `west spdx` script.

2. The output.
`west spdx` writes output as SPDX file and additionally this file is human-readable, but really not human-friendly. It contains a lot of unnecessary information and things that are actually important are lost among them. Our goal is to have user-friendly output, so we had to add new output format to upstream or rewrite current one to be more friendly.

3. The license detection.
`west spdx` has SPDX tag detection in the source code which is ok, but it does not cover all cases. We want to add external detector `Scancode-Toolkit` and some other method for `.a` libraries. (code for `.a` libraries is not in this PR yet)

Summarizing, adjusting upstream script requires rewriting most of the code and adding a lot of new, that's why decision was made to create our script.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+* You can read list of input files from a file:
+
+  .. code-block:: bash
+
+     --input-list-file list_file
```
```suggestion
  .. parsed-literal::
     :class: highlight

     --input-list-file *list_file*
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+Specifying output
+=================
+
```
Add some sort of an introductory text here before the bullets. For example something like:
You can specify the format of the report output using the ``output`` argument.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+* You can generate HTML report file:
+
+  .. code-block:: bash
+
+     --output-html <file-name.html>
```
```suggestion
  .. parsed-literal::
     :class: highlight

     --output-html *file-name.html*
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+Specifying output
+=================
+
+* You can generate HTML report file:
```
```suggestion
* To generate a report in HTML format:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+     --output-html <file-name.html>
+
+  If you use ``-d`` option, you can skip this option.
+  The ``sbom_report.html`` file will be generated in your build directory
+  (the first one if you specify more than one build directory).
```
```suggestion
  If you use ``-d`` option, you do not need to specify the report format.
  The :file:`sbom_report.html` file is generated in your build directory
  (the first one if you specify more than one build directory).
```

doki-nordic replied:
The user does not need to specify the output at all (not only `the report format`).

I am changing it to:

`If you use ``-d`` option, you do not need to specify any output argument.`

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+  The ``sbom_report.html`` file will be generated in your build directory
+  (the first one if you specify more than one build directory).
+
+* Special kind of output is a cache database:
```
```suggestion
* To generate a cache database:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+* Special kind of output is a cache database:
+
+  .. code-block:: bash
+
+     --output-cache-database <cache-database.json>
```
```suggestion
  .. parsed-literal::
     :class: highlight

     --output-cache-database *cache-database.json*
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+Detectors
+=========
+
+The SBOM command has following detectors implemented:
```
```suggestion
The ``sbom`` command has the following detectors implemented:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+* ``spdx-tag`` - search for the ``SPDX-License-Identifier`` in the source code or the binary file.
+  For guidelines, see `SPDX identifier`_. Enabled by default.
+
+* ``full-text`` - compare the contents of the source file with the small database of reference texts.
+  The database is part of the SBOM command. Enabled by default.
```
```suggestion
* ``full-text`` - compare the contents of the source file with a small database of reference texts.
  The database is part of the ``sbom`` command. Enabled by default.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+* ``scancode-toolkit`` - license detection by the `Scancode-Toolkit`_. Enabled and optional by default.
+
+  If ``scancode`` command is not on your ``PATH``,
+  you can use ``--scancode`` option to provide it, for example:
```
```suggestion
  If the ``scancode`` command is not on your ``PATH``, you can use the ``--scancode`` option to provide it, for example:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+* ``cache-database`` - use license information detected and cached earlier in the cache database file.
+  Disabled by default.
+
+  You have to provide cache database file with following argument:
```
```suggestion
  You have to provide the cache database file using the following argument:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+  You have to provide cache database file with following argument:
+
+  .. code-block:: bash
+
+     --input-cache-database <cache-database.json>
```
```suggestion
  .. parsed-literal::
     :class: highlight

     --input-cache-database *cache-database.json*
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+     --input-cache-database <cache-database.json>
+
+  Each database entry has a path relative to the west workspace directory, a hash and a list of detected licenses.
+  If the file under detection has the same path and hash, then the list of licenses from the database is used.
```
```suggestion
  Each database entry has a path relative to the west workspace directory, a hash, and a list of detected licenses.
  If the file under detection has the same path and hash, the list of licenses from the database is used.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+  If the file under detection has the same path and hash, then the list of licenses from the database is used.
+
+  .. note::
+     You can generate the database base on e.g scancode-toolkit detector by running following command:
```
```suggestion
     To generate the database based on, for example the scancode-toolkit detector, run the following command:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+  .. note::
+     You can generate the database base on e.g scancode-toolkit detector by running following command:
+
+     .. code-block:: bash
+
+        west sbom --input-files <files ...> --license-detectors scancode-toolkit --output-cache-database <cache-database.json>
```
```suggestion
     .. parsed-literal::
        :class: highlight

        west sbom --input-files *files ...* --license-detectors scancode-toolkit --output-cache-database *cache-database.json*
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+        west sbom --input-files <files ...> --license-detectors scancode-toolkit --output-cache-database <cache-database.json>
+
+If you prefer a non-default set of detectors,
+you can provide a list of comma separated detectors with the ``--license-detectors`` option, for example:
```
```suggestion
If you prefer a non-default set of detectors, you can provide a list of comma-separated detectors with the ``--license-detectors`` option, for example:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+  .. code-block:: bash
+
+     --license-detectors spdx-tag,scancode-toolkit
```
Does this need a space after the comma?

maje-emb replied:
The detector list should be separated only by comma.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+     --license-detectors spdx-tag,scancode-toolkit
+
+Some of the detectors can be optional which means that they are not executed for file that
+already has licenses detected by some other previously executed detector.
+Detectors are executed from left to right using list provided by the ``--license-detectors``.
```
```suggestion
Some of the detectors are optional, which means that they are not executed for a file that
already has licenses detected by some other previously executed detector.
Detectors are executed from left to right using a list provided by the ``--license-detectors``.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+     --optional-license-detectors scancode-toolkit
+
+To speed up detection time, some detectors may run in parallel on all available CPU cores.
+Use ``-n`` option to limit number of parallel threads or processes.
```
```suggestion
Some detectors may run in parallel on all available CPU cores, which speeds up the detection time.
Use the ``-n`` option to limit the number of parallel threads or processes.
```

maje-emb commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+     -d build_directory
+
+  The ``build_directory`` must be already built with the ``west build`` command using
+  ``Ninja`` as underlying build tool (which is default).
```
Should this sentence be split into two lines?

peknis replied:
> Should this sentence be split into two lines?

This is just one long sentence. I don't see a need for splitting it.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+Overview
+********
+
+The SBOM command can be divided into three basic stages:
```
```suggestion
The process of using the ``sbom`` command involves the following steps:
```

peknis commented:
Should we also add a bit of something in this part of NCS docs: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/introduction.html?highlight=spdx?

maje-emb commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,277 @@
+
+     --input-files file1 file2 ...
+
+  Each argument of this option can contain globs as defined by `Python's Path.glob`_ with two additions:
+  exclamation mark ``!`` to exclude files and absolute paths.
```
The second addition here is "absolute paths".
Maybe something tile this: `exclamation mark ``!`` to exclude files and support for an absolute paths.`?

peknis replied:
> The second addition here is "absolute paths". Maybe something tile this: `` exclamation mark ``!`` to exclude files and support for an absolute paths. ``?

Now that the first sentence says there are two additions, it is clearer if you have them as two separate bullets.

doki-nordic commented:
@peknis, please also review output HTML report. You are probably not familiar with Jinja2 HTML templates, so I created sample report in markdown that you can edit. I will later copy your changes to this PR.
https://github.com/doki-nordic/drafts/blob/west-sbom-report-draft-temp/west-sbom-report.md

mbolivar-nordic reviewed (changes requested):
Follow-up from internal discussion: please rename the extension to `ncs-sbom`.

Can you please include the information you gave on why the upstream `spdx` command is not sufficient in the commit log? I appreciate the explanation, but GitHub comments get lost easily. Git commit logs are forever, and are much easier to find when people are looking at the history of a file and trying to understand changes.

carlescufi commented:
> Follow-up from internal discussion: please rename the extension to `ncs-sbom`.
> 
> Can you please include the information you gave on why the upstream `spdx` command is not sufficient in the commit log? I appreciate the explanation, but GitHub comments get lost easily. Git commit logs are forever, and are much easier to find when people are looking at the history of a file and trying to understand changes.

+1 please be explicit in the commit logs so that someone with no previous knowledge can go back and understand the purpose of each of the commits without any additional sources of information. Thanks!

doki-nordic commented:
Sample report:
[sbom_report.zip](https://github.com/nrfconnect/sdk-nrf/files/8397734/sbom_report.zip)

maje-emb commented:
> If we need to leave the SPDX stuff in the docs:
> 
> * Is the new ncs-sbom tool an alternative for the SPDX? or
> * Is it something that can be used in addition to SPDX?
> * What is the added value of the new tool?

There is `west spdx` in upsteram. In addition, `ncs-sbom` was created, which is more extensive, with other features. 
See: https://github.com/nrfconnect/sdk-nrf/pull/7209#issuecomment-1082775563

doki-nordic commented:
Temporary switching to draft until we implement things that was agreed on the meeting.

doki-nordic commented:
Reopening the PR after adding commit [5dc02a1](https://github.com/nrfconnect/sdk-nrf/pull/7209/commits/5dc02a164de96194e2b266b543cea5500d56cdeb) with the changes that ware recently requested:
* Documentation:
   * `west ncs-sbom` mentioned in the introduction page
   * Added general overview what user can find the the HTML report.
   * Description how we are generating the list of files from the application build (from user perspective).
   * Added note that build must be successful and up to date, i.e. you have to rebuild it on each change of kconfig option.
* command renamed to `west ncs-sbom`
* Added SPDX output format
* Significant improvements in extracting list of files from an application build including verification of the list based on `.map` file parsing.
* Improvements in scancode-toolkit detector, now it produces more SPDX compatible results and adds license user-friendly name and URL if available.
* Added "External file detector" that gets license tag from an external file, see: https://github.com/nrfconnect/sdk-nrfxlib/pull/712
* Non-standard license identifiers ranamed, so now they don't have GUID part, but tool prefix (like in scancode-toolkit), e.g. `LicenseRef-BSD-3-Clause-Intel-F438A16F-DDC6-41CC-87BE-F1BE8A937F84` -> `LicenseRef-west-ncs-sbom-BSD-3-Clause-Intel`
* HTML output style improvements:
   * By default, the same style as the NCS documentation.
   * If opened in a browser setup to show pages in dark mode, it will switch to a dark mode.
   * If opened in Visual Studio Code, it will use its color theme.
* Added note to the report that some licenses were detected by scancode-toolkit with link to it

peknis commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -410,6 +410,8 @@ Scripts
 
 This section provides detailed lists of changes by :ref:`script <scripts>`.
 
+* Added new ``west sbom`` command that generates :ref:`Software Bill of Materials <west_sbom>`.
```
```suggestion
* Added new ``west ncs-sbom`` command that generates :ref:`Software Bill of Materials <west_sbom>`.
```

peknis commented on doc/nrf/introduction.rst:
```diff
@@ -35,8 +35,12 @@ Each file included in the repositories also has an `SPDX identifier`_ that menti
 
-The `SPDX tool`_ is used to generate license reports on each release of the |NCS|.
-You can also use SPDX to generate license reports for your projects that are specific to the code included in your application.
+The west :ref:`ncs-sbom <west_sbom>` utility is used to generate a license report.
```
```suggestion
You can use the west :ref:`ncs-sbom <west_sbom>` utility to generate a license report.
```

peknis commented on doc/nrf/introduction.rst:
```diff
@@ -35,8 +35,12 @@ Each file included in the repositories also has an `SPDX identifier`_ that menti
-The `SPDX tool`_ is used to generate license reports on each release of the |NCS|.
-You can also use SPDX to generate license reports for your projects that are specific to the code included in your application.
+The west :ref:`ncs-sbom <west_sbom>` utility is used to generate a license report.
+It allows to generate report for the |NCS|, built application or for specific files.
```
```suggestion
It allows you to generate a report for the |NCS|, built application, or specific files.
```

peknis commented on doc/nrf/introduction.rst:
```diff
@@ -35,8 +35,12 @@ Each file included in the repositories also has an `SPDX identifier`_ that menti
-You can also use SPDX to generate license reports for your projects that are specific to the code included in your application.
+The west :ref:`ncs-sbom <west_sbom>` utility is used to generate a license report.
+It allows to generate report for the |NCS|, built application or for specific files.
+The tool is configurable in a wide range.
```
```suggestion
The tool is highly configurable.
```

peknis commented on doc/nrf/introduction.rst:
```diff
@@ -35,8 +35,12 @@ Each file included in the repositories also has an `SPDX identifier`_ that menti
+The west :ref:`ncs-sbom <west_sbom>` utility is used to generate a license report.
+It allows to generate report for the |NCS|, built application or for specific files.
+The tool is configurable in a wide range.
+It uses several detection methods e.g. based on spdx tags, searching license information in files or using the `Scancode-Toolkit`_.
```
```suggestion
It uses several detection methods, such as:

 * Search based on SPDX tags.
 * Search license information in files.
 * Use of the `Scancode-Toolkit`_.
```

peknis commented on doc/nrf/introduction.rst:
```diff
@@ -35,8 +35,12 @@ Each file included in the repositories also has an `SPDX identifier`_ that menti
+It allows to generate report for the |NCS|, built application or for specific files.
+The tool is configurable in a wide range.
+It uses several detection methods e.g. based on spdx tags, searching license information in files or using the `Scancode-Toolkit`_.
+The report is generated in HTML and/or SPDX format.
```
Depending on your configuration, the report is generated in HTML or SPDX format, or in both formats.

peknis replied:
```suggestion
Depending on your configuration, the report is generated in HTML or SPDX format, or in both formats.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,367 @@
+    Generating list of licenses from an application build is experimental.
+    The accuracy of detection is constantly verified.
+    Both implementation and usage may change in the future.
+
```
Remove extra line.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,367 @@
+#. Create output report containing all the files and license information related to them,
+   for example, write a report file in HTML format.
+   For details, see :ref:`west_sbom Specifying output`.
+
```
Remove extra line.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,367 @@
+
+    For more details, see `Scancode-Toolkit Installation`_.
+
+
```
Remove extra line.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,367 @@
+
+     west ncs-sbom --input-files *file1* *file2* --output-html *file-name.html*
+
+
```
Remove extra line.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,367 @@
+You can specify all input options several times to provide more input for the report generation, for example, generate a report for two applications.
+You can also mix them, for example, to generate a report for the application and some directory.
+
+
```
Remove extra line.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,367 @@
+
+  For details, see ``cache-database`` detector.
+
+
```
Remove extra line.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,367 @@
+     :class: highlight
+
+     -d build_directory *target.elf*:*file.map*
+
```
Remove extra line.

mbolivar-nordic commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -633,6 +633,8 @@ Scripts
 
 This section provides detailed lists of changes by :ref:`script <scripts>`.
 
+* Added new ``west sbom`` command that generates :ref:`Software bill of materials <sbom_script>`.
```
Various places in this PR (including documentation, e.g. here, but also the commit log for this patch) are still referring to `west sbom` instead of `west ncs-sbom`.

The explanation of why this is needed when `west spdx` is already present is also not present in the commit log (or the documentation, as far as I can see) as discussed.

doki-nordic replied:
The new commit message will be:

```
scripts: sbom: Add "west sbom" tool.

The "west ncs-sbom" tool creates a Software Bill of Materials.
It uses a built application or the specific files as an input,
detects their licenses, and creates an output report that can
be in HTML or SPDX format.

It is significantly different than the Zephyr's "west spdx"
tool, so it was implemented from the. It has a different
method of getting files from the application that gets a more
precise list of files. It has also additional methods for
verifying that the list is correct. It adds more license
detection methods including an external detector from
"scancode-toolkit". The output (both HTML and SPDX) is more
readable.
```

miha-nordic commented on scripts/west_commands/sbom/license_utils.py:
```diff
@@ -0,0 +1,140 @@
+    or_present: bool = False
+
+
+def get_spdx_license_expr_info(expr: str) -> SPDXLicenseExprInfo:
```
Are those class methods or just standalone functions?

doki-nordic replied:
The intention was standalone function that returns multiple values. I used class to enclose them in a single, well defined container.

doki-nordic commented:
@peknis, I haven't seen your review of the `HTML report overview` and `Extracting list of files from a build directory` chapters. They were added after this PR exited "Draft" state again, so maybe you missed that. I don't believe that I wrote such a long text without any mistake ;)

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+* ``external-file`` - search for license information in an external file. Enabled by default.
+
+  The external file has the following properties:
```
```suggestion
  This detector has the following properties:
```

doki-nordic replied:
Not `detector`, but the external file that is used by this detector. The `external file` was mentioned in above sentence, so I think it is pretty clear. I would not change this.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+  The external file has the following properties:
+
+    * It is located in the same directory or any of the parent directories of the file under detection.
```
```suggestion
    * It is located in the same directory as the file under detection or in one of its parent directories .
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+    * It is located in the same directory or any of the parent directories of the file under detection.
+    * Its name contains ``LICENSE``, ``LICENCE`` or ``COPYING`` (case insensitive).
+    * It has a ``SPDX-License-Identifier`` tag.
```
```suggestion
    * It has an ``SPDX-License-Identifier`` tag.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+    * It is located in the same directory or any of the parent directories of the file under detection.
+    * Its name contains ``LICENSE``, ``LICENCE`` or ``COPYING`` (case insensitive).
+    * It has a ``SPDX-License-Identifier`` tag.
+    * It has one or more ``NCS-SBOM-Apply-To-File`` tags containing file paths or globs (as defined by the `Python's Path.glob`_).
```
```suggestion
    * It has one or several ``NCS-SBOM-Apply-To-File`` tags containing file paths or globs (as defined by the `Python's Path.glob`_).
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+* ``cache-database`` - use license information detected and cached earlier in the cache database file.
+  Disabled by default.
+
+  You have to provide the cache database file using the following argument:
```
```suggestion
  Provide the cache database file using the following argument:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+  If the file under detection has the same path and hash, the list of licenses from the database is used.
+
+  .. note::
+     To generate the database based on, for example the scancode-toolkit detector, run the following command:
```
```suggestion
     To generate the database based on, for example the ``scancode-toolkit`` detector, run the following command:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+The HTML report has following structure:
+
+* Summary of the report. It contains:
```
```suggestion
* Summary of the report, containing the following:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+* Summary of the report. It contains:
+
+   * At the beginning are notes.
+     They will give you some general information that you should pay attention while reading the report.
```
```suggestion
   * Notes at the beginning.

     General information on the report.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+   * At the beginning are notes.
+     They will give you some general information that you should pay attention while reading the report.
+   * The list of inputs.
+     It shows from where this report takes the files.
```
```suggestion
   * List of inputs.

     The file sources.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+     They will give you some general information that you should pay attention while reading the report.
+   * The list of inputs.
+     It shows from where this report takes the files.
+   * The list of licenses.
+     It contains all licenses detected in the input files.
```
```suggestion
   * List of licenses.

     All licenses detected in the input files.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+     It shows from where this report takes the files.
+   * The list of licenses.
+     It contains all licenses detected in the input files.
+   * The list of added license texts.
+     If a license is not in the `SPDX License List`_ and we have it in our internal database,
+     the license text will be added to the report.
```
```suggestion
   * List of added license texts.

     If a license is not in the `SPDX License List`_ and it is in the internal database,
     the license text is added to the report.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+     If a license is not in the `SPDX License List`_ and we have it in our internal database,
+     the license text will be added to the report.
+
+  You can click links in the summary to get more details about specific item.
```
```suggestion
  You can click links in the summary to get more details about specific items.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+  You can click links in the summary to get more details about specific item.
+
+* List of files without any license information or with license information that cannot be detected automatically.
+  You have to investigate them manually to get the license information.
```
```suggestion
* List of files without any license information or with license information that cannot be detected automatically.

  You have to investigate them manually to get the license information.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+* Details about each detected license:
+
+   * License identifier
+   * Information if it is a standard SPDX license
+   * License name if available
+   * Link to license text or more details if available
+   * All files from the input covered by this license
```
```suggestion
   * License identifier.
   * Information if it is a standard SPDX license.
   * License name if available.
   * Link to license text or more details if available.
   * All files from the input covered by this license.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+.. _west_sbom Extracting from build:
+
+Extracting list of files from a build directory
+***********************************************
```
```suggestion
Extracting a list of files from a build directory
*************************************************
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+Extracting list of files from a build directory
+***********************************************
+
+The ``ncs-sbom`` extracts a list of files from a build directory by querying the ``ninja`` about its targets and dependencies.
```
```suggestion
The ``ncs-sbom`` extracts a list of files from a build directory.
It queries ``ninja`` for the targets and dependencies.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+The ``ncs-sbom`` extracts a list of files from a build directory by querying the ``ninja`` about its targets and dependencies.
+
+The entry point is the ``zephyr/zephyr.elf`` target.
```
```suggestion
The entry point is the :file:`zephyr/zephyr.elf` target file.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+The ``ncs-sbom`` extracts a list of files from a build directory by querying the ``ninja`` about its targets and dependencies.
+
+The entry point is the ``zephyr/zephyr.elf`` target.
+The script asks ``ninja`` for all input targets of the ``zephyr/zephyr.elf`` target.
```
```suggestion
The script asks ``ninja`` for all input targets of the :file:`zephyr/zephyr.elf` target.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+The entry point is the ``zephyr/zephyr.elf`` target.
+The script asks ``ninja`` for all input targets of the ``zephyr/zephyr.elf`` target.
+Next, it asks for all input targets of the previously extracted input targets,
+and so on until it reaches all leaves in the dependency tree.
```
```suggestion
It also asks for all input targets of the previously extracted input targets,
until it reaches all leaves in the dependency tree.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+and so on until it reaches all leaves in the dependency tree.
+The result is a list of all the leaves.
+
+You can change the target or specify multiple targets by adding them after the build directory in the ``-d`` option, for example:
```
```suggestion
To change the target or specify multiple targets, you can add them after the build directory in the ``-d`` option, for example:
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+   -d build_directory *target1.elf* *target2.elf*
+
+Two redundant methods for increasing the correctness of the above algorithm are implemented:
```
```suggestion
There are two redundant methods for improving the correctness of the above algorithm:
``

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+
+Two redundant methods for increasing the correctness of the above algorithm are implemented:
+
+* Each library is examined using the GNU ``ar`` tool.
+  If the list of files returned by the GNU ``ar`` tool is covered by the list returned from the ``ninja``,
+  the list is assumed to be valid.
+  Otherwise, the library is assumed to be a leaf, so it is shown in the report and its inputs are not analyzed further.
```
```suggestion
* Each library is examined using the GNU ``ar`` tool.

  If the list of files returned by the GNU ``ar`` tool is covered by the list returned from the ``ninja``, the list is assumed to be valid.
  Otherwise, the library is assumed to be a leaf, so it is shown in the report and its inputs are not analyzed further.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,361 @@
+  the list is assumed to be valid.
+  Otherwise, the library is assumed to be a leaf, so it is shown in the report and its inputs are not analyzed further.
+
+* The ``ncs-sbom`` pareses the :file:`.map` file created during the ``zephyr/zephyr.elf`` linking.
+  It gives a list of all object files and libraries linked into ``zephyr/zephyr.elf``.
+  The script ends with a fatal error if some file in the :file:`.map` file is not visible by the ``ninja``.
```
```suggestion
* The ``ncs-sbom`` pareses the :file:`.map` file created during the :file:`zephyr/zephyr.elf` linking.

  It gives a list of all object files and libraries linked to the :file:`zephyr/zephyr.elf` file.
  The script ends with a fatal error if any file in the :file:`.map` file is not visible by ``ninja``.
```

doki-nordic commented:
@peknis, I added all your changes except one: https://github.com/nrfconnect/sdk-nrf/pull/7209#discussion_r875623065

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,370 @@
+
+The ``ncs-sbom`` command requires installation of additional Python packages.
+
+Use the following command to install the requirements.
```
```suggestion
Use a command appropriate for your operating system to install the requirements.
```

doki-nordic replied:
The commands are shown in a tabs for each operating system. For developer, it is pretty obvious what he should do, so in my opinion stating the obvious is not needed here, but if you think that it is better that way, then ok.

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,370 @@
+   :depth: 2
+
+The Software Bill of Materials (SBOM) is a :ref:`west <zephyr:west>` extension command that can be invoked by ``west ncs-sbom``.
+It provides a list of used licenses for an application build or the specific files.
```
```suggestion
It provides a list of used licenses for an application build or specific files.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,370 @@
+   -d build_directory *target1.elf* *target2.elf*
+
+There are two redundant methods for improving the correctness of the above algorithm:
+``
```
```suggestion
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,370 @@
+
+   -d build_directory *target1.elf* *target2.elf*
+
+There are two redundant methods for improving the correctness of the above algorithm:
```
```suggestion
There are two alternative methods for improving the correctness of the above algorithm:
```

doki-nordic replied:
In this context `redundancy` does not mean the same as `alternative`.

peknis replied:
Okay. How about "additional"?

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,370 @@
+  If the list of files returned by the GNU ``ar`` tool is covered by the list returned from the ``ninja``, the list is assumed to be valid.
+  Otherwise, the library is assumed to be a leaf, so it is shown in the report and its inputs are not analyzed further.
+
+* The ``ncs-sbom`` pareses the :file:`.map` file created during the :file:`zephyr/zephyr.elf` linking.
```
```suggestion
* The ``ncs-sbom`` parses the :file:`.map` file created during the :file:`zephyr/zephyr.elf` linking.
```

peknis commented on scripts/west_commands/sbom/README.rst:
```diff
@@ -0,0 +1,370 @@
+
+* The ``ncs-sbom`` pareses the :file:`.map` file created during the :file:`zephyr/zephyr.elf` linking.
+
+  It gives a list of all object files and libraries linked to the :file:`zephyr/zephyr.elf` file.
```
```suggestion
  It provides a list of all object files and libraries linked to the :file:`zephyr/zephyr.elf` file.
```

doki-nordic commented:
I rebased and squashed it to make this PR ready for merging.

If someone is interested in the PR history: https://github.com/doki-nordic/fw-nrfconnect-nrf/tree/license-gen-before-squash

carlescufi commented on west.yml:
```diff
@@ -182,6 +182,5 @@ manifest:
     path: nrf
-    # This line configures west extensions which are currently only
-    # for internal use by NCS maintainers.
+    # This line configures west extensions.
```
@doki-nordic do we need this change?
